### PR TITLE
Implemented reading of routes via SCR agents (dropped Yast::Routing.Read)

### PR DESCRIPTION
**Pull request by mchf**

### CODE REVIEW

File: src/lib/y2network/config_reader/routes_reader.rb [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+          end
+
+          route
+        end
+      end
+
+      # Given an IP and a netmaks, returns a valid IPAddr objecto
+      #
+      # @param ip_str      [String] IP address; {MISSING_VALUE} means that the IP is not defined
+      # @param netmask_str [String] Netmask; {MISSING_VALUE} means than no netmaks was specified
```
Comment (teclator) [inline src/lib/y2network/config_reader/routes_reader.rb]:
```suggestion
      # @param netmask_str [String] Netmask; {MISSING_VALUE} means that no netmask was specified
```

File: src/lib/y2network/config_reader/routes_reader.rb [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+
+require "yast"
+
+module Y2Network
+  module ConfigReader
+    DEFAULT_ROUTES_FILE = "/etc/sysconfig/network/routes".freeze
+
+    # This class reads the current configuration from a file in routes format
+    # (@see man routes)
+    class RoutesReader
```
[the remark this answers is not in the document]
Comment (mchf) [reply src/lib/y2network/config_reader/routes_reader.rb]: in fact i was thinking of this name when implementing ... and it was a bit long for me ... what about something like ScrRoutesReader?

File: src/lib/y2network/config_reader/routes_reader.rb [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+      #
+      # @param routes [Array<Hash>] in quad or CIDR flavors (see {#Routes})
+      # @return [Array<Hash>] in quad or slash flavor
+      def normalize_routes(routes)
+        return routes if routes.nil? || routes.empty?
+
+        routes.map do |route|
+          subnet, prefix = route["destination"].split("/")
+
+          if !prefix.nil?
```
[the remark this answers is not in the document]
Comment (mchf) [reply src/lib/y2network/config_reader/routes_reader.rb]: well it is copy pasted method from old Routing module ... but I can clearly do something with that

File: src/lib/y2network/config_reader/routes_reader.rb [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+          route
+        end
+      end
+
+      # Given an IP and a netmaks, returns a valid IPAddr objecto
+      #
+      # @param ip_str      [String] IP address; {MISSING_VALUE} means that the IP is not defined
+      # @param netmask_str [String] Netmask; {MISSING_VALUE} means than no netmaks was specified
+      # @return [IPAddr,nil] The IP address or `nil` if the IP is missing
+      def build_ip(ip_str, netmask_str = nil)
```
[the remark this answers is not in the document]
Comment (mchf) [reply src/lib/y2network/config_reader/routes_reader.rb]: yep, I usually rebase ... and next time clearly will again ;-)

File: src/lib/y2network/config_reader/sysconfig.rb [hunk 5]
```diff
@@ -29,18 +31,17 @@ module ConfigReader
...
+        Config.new(interfaces: interfaces, routing: routing, source: :sysconfig)
       end
 
     private
 
-      MISSING_VALUE = "-".freeze
-      private_constant :MISSING_VALUE
-
-      # Find network interfaces
+      # Find configured network interfaces
```
[the remark this answers is not in the document]
Comment (mchf) [reply src/lib/y2network/config_reader/sysconfig.rb]: interfaces that has a configuration (ifcfg file) assigned ... you can also have an interface without configuration (e.g. an unused NIC card) ... LanItems works even with such devices / interfaces but from routing POV these are useless.

[the remark this answers is not in the document]
Comment (imobachgs) [reply src/lib/y2network/config_reader/sysconfig.rb]: Could you just write a comment there? So I won't make this question again :)

File: src/lib/y2network/config_reader/routes_reader.rb [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+          if !prefix.nil?
+            route["destination"] = subnet
+            route["netmask"] = "/#{prefix}"
+          end
+
+          route
+        end
+      end
+
+      # Given an IP and a netmaks, returns a valid IPAddr objecto
```
[the remark this answers is not in the document]
Comment (mchf) [reply src/lib/y2network/config_reader/routes_reader.rb]: ok

File: src/lib/y2network/config.rb [hunk 7]
```diff
@@ -30,48 +30,38 @@ module Y2Network
...
 
+    # @param source [Symbol] Source to read the configuration from
     class << self
-      # @param source [Symbol] Source to read the configuration from
-      # @return [Y2Network::Config]
       def from(source)
-        reader = ConfigReader.for(source)
-        reader.config
+        builder = ConfigReader.for(source)
+        builder.network_config
```
[the remark this answers is not in the document]
Comment (mchf) [reply src/lib/y2network/config.rb]: ok

File: src/lib/y2network/config.rb [hunk 8]
```diff
@@ -30,48 +30,38 @@ module Y2Network
...
       end
     end
 
     # Constructor
     #
     # @param id             [Symbol] Configuration ID
     # @param interfaces     [Array<Interface>] List of interfaces
-    # @param routing_tables [Array<RoutingTable>] List of routing tables
-    def initialize(id: :system, interfaces:, routing_tables:, source:)
+    # @param routing        [Routing] Object with routing configuration
```
[the remark this answers is not in the document]
Comment (mchf) [reply src/lib/y2network/config.rb]: ok

File: src/lib/y2network/config_reader/routes_reader.rb (hunk 4 above)
[the remark this answers is not in the document]
Comment (mchf) [reply src/lib/y2network/config_reader/routes_reader.rb]: ok

File: src/lib/y2network/config_reader/sysconfig.rb [hunk 9]
```diff
@@ -29,18 +31,17 @@ module ConfigReader
     # This class reads the current configuration from `/etc/sysconfig` files
     class Sysconfig
       # @return [Y2Network::Config] Network configuration
-      def config
+      def network_config
```
[the remark this answers is not in the document]
Comment (mchf) [reply src/lib/y2network/config_reader/sysconfig.rb]: ok

File: src/lib/y2network/config_reader/sysconfig.rb (hunk 5 above)
[the remark this answers is not in the document]
Comment (mchf) [reply src/lib/y2network/config_reader/sysconfig.rb]: ok

File: src/lib/y2network/config_reader/sysconfig.rb [hunk 10]
```diff
@@ -29,18 +31,17 @@ module ConfigReader
...
         interfaces = find_interfaces
-        routing_tables = find_routing_tables(interfaces)
-        Config.new(interfaces: interfaces, routing_tables: routing_tables, source: :sysconfig)
+        # load /etc/sysconfig/network/routes
+        routing = Y2Network::Routing.new(tables: [RoutesReader.new.config])
```
[the remark this answers is not in the document]
Comment (mchf) [reply src/lib/y2network/config_reader/sysconfig.rb]: ok

File: src/lib/y2network/config_reader/routes_reader.rb (hunk 3 above)
[the remark this answers is not in the document]
Comment (mchf) [reply src/lib/y2network/config_reader/routes_reader.rb]: ok

File: src/lib/y2network/config_reader/routes_reader.rb (hunk 1 above)
Comment (mchf) [reply src/lib/y2network/config_reader/routes_reader.rb]: ok

File: src/lib/y2network/config_reader/routes_reader.rb (hunk 2 above)
[the remark this answers is not in the document]
Comment (mchf) [reply src/lib/y2network/config_reader/routes_reader.rb]: Scr sticks too much with access method -> using Sysconfig

File: src/lib/y2network/config_reader/sysconfig_routes_reader.rb [hunk 11]
```diff
@@ -0,0 +1,121 @@
...
+        netmask_str == MISSING_VALUE ? ip : ip.mask(netmask_str)
+      end
+
+      # Build a route given a hash from the SCR agent
+      #
+      # @param hash [Hash] Hash from the `.routes` SCR agent
+      # @return Route
+      def build_route(hash)
+        # TODO: check whether the iface is configured in the system
+        iface = Interface.new(hash["device"])
```
Comment (imobachgs) [inline src/lib/y2network/config_reader/sysconfig_routes_reader.rb]: Please, require `y2network/interface` above.

File: src/lib/y2network/config_reader/sysconfig_routes_reader.rb [hunk 12]
```diff
@@ -0,0 +1,121 @@
...
+  module ConfigReader
+    DEFAULT_ROUTES_FILE = "/etc/sysconfig/network/routes".freeze
+
+    # This class reads the current configuration from a file in routes format
+    # (@see man routes)
+    class SysconfigRoutesReader
+      # @param routes_file [<String>] full path to a file in routes format, when
+      #                               not defined, then /etc/sysconfig/network/routes is used
+      def initialize(routes_file: DEFAULT_ROUTES_FILE)
+        @routes_file = Yast::Path.new(".routes") if routes_file == DEFAULT_ROUTES_FILE
```
[the remark this answers is not in the document]
Comment (mchf) [reply src/lib/y2network/config_reader/sysconfig_routes_reader.rb]: will continue soon - prepartion for ifroute files reading.

[the remark this answers is not in the document]
Comment (mchf) [reply src/lib/y2network/config_reader/sysconfig_routes_reader.rb]: (this part "if" escaped me from another branch)

File: src/lib/y2network/config_reader/sysconfig_routes_reader.rb [hunk 13]
```diff
@@ -0,0 +1,122 @@
...
+        iface = Interface.new(hash["device"])
+        # normalized SCR output contains either subnet mask or /<prefix length> under
+        # "netmask" key
+        # TODO: this should be improved in normalize_routes
+        mask = hash["netmask"] =~ /\/[0-9]+/ ? hash["netmask"][1..-1] : hash["netmask"]
+
+        Y2Network::Route.new(
+          to:        hash["destination"] != DEFAULT_DEST ? build_ip(hash["destination"], mask) : :default,
+          interface: iface,
+          gateway:   build_ip(hash["gateway"], MISSING_VALUE)
```
Comment (teclator) [inline src/lib/y2network/config_reader/sysconfig_routes_reader.rb]: options is not imported

File: src/lib/y2network/config_writer/sysconfig.rb [hunk 14]
```diff
@@ -31,7 +31,7 @@ class Sysconfig
       #
       # @param config [Y2Network::Config] Configuration to write
       def write(config)
-        routes = config.routes.map { |r| route_to_hash(r) }
+        routes = config.routing.routes.map { |r| route_to_hash(r) }
```
Comment (teclator) [inline src/lib/y2network/config_writer/sysconfig.rb]: route_to_hash writes default route as "-" but we only support import with "default"

File: src/lib/y2network/config_reader/sysconfig_routes_reader.rb (hunk 13 above)
Comment (mchf) [reply src/lib/y2network/config_reader/sysconfig_routes_reader.rb]: will fix when merging the ifroute PR (need to fix its agent which ignores options anyway)

File: src/lib/y2network/config_writer/sysconfig.rb (hunk 14 above)
Comment (mchf) [reply src/lib/y2network/config_writer/sysconfig.rb]: will fix when merging scr write PR

**Pull request merged**